### ISSUE TITLE
Custom Piano Labels

### ISSUE THREAD

**Issue by anytizer** [LMMS/lmms#6162]
English | Best Alternative | Roman short form | North Indian | Sanskrit | South Indian
-- | -- | -- | -- | -- | --
C | Sa | S | सा | षड्ज | सा
D | Re | R | रे | ऋषभ | री
E | Ga | G | ग | गान्धार | गा
F | Ma | M | म | मध्यम | म
G | Pa | P | प | पंचम | प
A | Dha | D | ध | धैवत | ध
B | Ni | N | नि | निषाद | नि

It would be fantastic if one can change the piano labels to custom labels as used in many other practices.
The table above shows some basic replacements (what to replace with).

![image](https://user-images.githubusercontent.com/5563341/133643210-660907bd-29b2-4c17-ab83-a0e75a8342dc.png)

Many authors (tutors in particular) use their notes for comfort. But while it comes to data-entry the existing notes in LMMS; it is confusing to see English Labels only. It would help a beginner do the data entry so easily, if the custom labels were possible.

Calc/Spreadsheet File attached:
[custom-labels.ods](https://github.com/LMMS/lmms/files/7179228/custom-labels.ods)

**Comment by Veratil** [comment]
A couple followup questions:
What about sharps and flats?
Would the octave number need to be changed as well? I'm assuming yes.

**Comment by anytizer** [comment]
Sharp, Flats and Octave numbers representation should remain unchanged for now (which is still fine).

In some cases:
- C4 to B4 can be written without numbers. eg. D
- C5 - B5 can have a dot on the top side ( * ) eg: D*.
- C3 - B3 can have an underscore or a lower dot ( . ) eg, D.
Hence, D., D, D* will represent keys from three different octaves: mandhra saptak, madhya saptak and taar saptak.
(saptak = a full octave note not including the first/last key of another octave), probably meaning group of 7.
mandhra = lower octave name
madhya = mid octave name
taar = higher octave name

When necessary to use note from another octave, a double symbol may be used. eg. D** for D6.

An example of notations is in the description of [this video](https://www.youtube.com/watch?v=nDkZ_vjgD-U).

**Comment by he29-net** [comment]
Just as an idea for implementation, this could (should?) be integrated to the new tuning system. Simply save an identifier for each scale degree along with its ratio / interval. Just as `getFrequency(key)` returns the pitch based on given scale degree, `getLabel(key)` could return whatever string the user set for the scale degree (probably combined with the octave number). Changing the set of labels would be then as simple as switching between scales (or overwriting the default scale). Not to mention that if the user needs Indian labels for example, there is a good chance they may want to use an Indian scale / tuning system. So this should be a natural fit for such functionality.

One problem is that the Scala tuning format does not support this. To work around it, I would suggest adding another "LMMS extension" to the format, similar to the one that is used in our keymap files. I think I used a double comment character (`;;`) to identify a line that contains name of the keymap (since name is not present in the original Scala keymap format). Scala and other software would see it as an ordinary comment, so the format stays 100 % compatible with the original.

Similarly, a "special comment" added after a cent or fraction value in the `.scl` file could define a label for that scale degree. Unfortunately, ZynAddSubFX (and possibly other SW) is broken in that regard: the Scala specification says that everything that appears on the line after a cent or fraction value (i.e. comments) should be ignored, but Zyn will fail to correctly load a file with such comments. Alternatively, the optional "special comment" could be expected on the next line, after the scale degree. That would be not as nice, but probably more compatible with broken software.

**Comment by Spekular** [comment]
I definitely agree that this should be mappable if added. My honest opinion is that adding alternative note labels is a bit like putting training wheels on training wheels, but if we're doing it then it shouldn't be hard coded.

Also, if we're bundling the labels proposed above then Do Re Mi Fa So La Ti Do and CDEFGHB should be included as well.

**Comment by anytizer** [comment]
That's true about mapping, so that we could add any kind of notation practice.
(PS: @Spekular I updated 2 columns in the primary comment for reference).

Currently, there should be very little change to the code (not wanting to introduce any new bugs due to this change).
So, we should pick an easiest and minimal way for mapping.

Also, does LMMS use .sqlite somewhere to read/write configs?

**Comment by Veratil** [comment]
> Also, does LMMS use .sqlite somewhere to read/write configs?

No, the save format is XML.

**Comment by anytizer** [comment]
Not sure if the below xml helps. Also, it needs to work with the Unicode characters as well.

	<?xml version="1.0" encoding="UTF-8" standalone="no" ?>
	<scale name="Indian Classical - Romanized">
		<label name="C" notation="Sa">
		<label name="D" notation="Re">
		<label name="E" notation="Ga">
		<label name="F" notation="Ma">
		<label name="G" notation="Pa">
		<label name="A" notation="Dha">
		<label name="B" notation="Ni">
	</scale>
	<scale name="Indian Classical - Unicode">
		<label name="C" notation="सा">
		<label name="D" notation="रे">
		<label name="E" notation="ग">
		<label name="F" notation="म">
		<label name="G" notation="प">
		<label name="A" notation="ध">
		<label name="B" notation="नि">
	</scale>
	<scale name="Solfège - Western">
		<label name="C" notation="Do">
		<label name="D" notation="Re">
		<label name="E" notation="Mi">
		<label name="F" notation="Fa">
		<label name="G" notation="So">
		<label name="A" notation="La">
		<label name="B" notation="Ti">
	</scale>
	<scale name="German">
		<label name="C" notation="C">
		<label name="D" notation="D">
		<label name="E" notation="E">
		<label name="F" notation="F">
		<label name="G" notation="G">
		<label name="A" notation="A">
		<label name="B" notation="H">
	</scale>
	<scale name="English">
		<label name="C" notation="C">
		<label name="D" notation="D">
		<label name="E" notation="E">
		<label name="F" notation="F">
		<label name="G" notation="G">
		<label name="A" notation="A">
		<label name="B" notation="B">
	</scale>

2 ways to deal:
- Put the xml in /configs/scales/*.xml
- Or, make a single file appending `<scale>` block for each config.

It should be ok to repeat `<label name="..."` in each xml, because this is what can be used to return the notation.

**Comment by Monospace-V** [comment]
I believe there is something called drum-mapping, which some DAWs support. This enhancement request could help map drums too.

**Comment by PhysSong** [comment]
> I believe there is something called drum-mapping, which some DAWs support. This enhancement request could help map drums too.

Related: #5357

**Comment by anytizer** [comment]
**External solution:** I created a temporary solution: [swar-converter application](https://github.com/anytizer/swar.lmms/blob/main/swar/libraries/XMLHandler.cs) that can generate .xpt files. These files can later be imported in LMMS piano editor. Please see how it works. Basically, it converts sargam notes into English Scales and meanwhile, produces an .xpt file.

Source written in C# using Visual Studio. Compilation gives you a Windows Binary file. Rest of the details are in the application itself. Just linked in here, if it helps.

**Comment by anytizer** [comment]
One of the well formatted notations for reference is [here](https://github.com/anytizer/melodies.lmms/blob/main/melodies/%E0%A4%A4%E0%A5%80%E0%A4%9C%20%E0%A4%AD%E0%A4%BE%E0%A4%95%E0%A4%BE%20-%20%E0%A4%AA%E0%A5%81%E0%A4%B0%E0%A4%BE%E0%A4%A8%E0%A5%8B/notations/notations-sargams.txt).

**Comment by anytizer** [comment]
Some findings:

Unicode support required in the labels.
![Unicode support required](https://user-images.githubusercontent.com/5563341/218281570-e25d87d1-e65d-40ea-96d0-2a3b04d55504.png)

Space for writing the labels over piano drawing needs often 4 - 5 letters. Otherwise, it is not legible to read at 100%.
![keymap-200-percent](https://user-images.githubusercontent.com/5563341/218281566-0bbaefbf-1db4-4ab2-bb33-e7b3580de224.png)

https://github.com/LMMS/lmms/blob/507fa239ddffd2ed1f186d0a51016c85d9c2b5a9/src/gui/editors/PianoRoll.cpp#L132

**Comment by anytizer** [comment]
Basic process is a 3 step change in PianoRoll.cpp file.

## Step 1:

Obtain your own key/note names somewhere around line 130.
```
static std::array<QString, 12> s_noteStrings {"S", "r", "R", "g", "G", "m", "M", "P", "d", "D", "n", "N"};
```
These values may be read from translation or config files.

## Step 2:
Add a parameter to control the prefix and suffixes for octave number. Replace the function as:
```
static QString getNoteString(int key, bool showfixes=false)
{
	QString prefix = "";
	QString suffix = "";
	
	int octave = static_cast<int>(FirstOctave + key / KeysPerOctave);
	
	if(showfixes==true)
	{
		if(octave<4) {for(int l=3; l>=octave; --l){prefix += "."; }}
		if(octave>4) {for(int h=5; h<=octave; ++h){suffix += "*"; }}
	}
	return prefix + s_noteStrings[key % 12] + suffix;
}
```

## Step 3:
```
Line near by 1070: // show octave notice in actual tone (green marker)
QString noteKeyString = getNoteString(n->key(), true);

Line near by 3262: // hide octave notice in the piano key (while kyes)
QString noteString = getNoteString(key, false);
```

**Comment by anytizer** [comment]
List of some related changes:
#6865 - Added flats to array of keys
